File: auditor.py
from decimal import Decimal
from schwifty import IBAN
from schemas import NoticeExtraction

def audit_notice(data: NoticeExtraction, pre_approved_mandate: dict | None = None) -> dict:
    audit_results = {
        "math_passed": True,
        "wire_passed": True,
        "errors": [],
        "warnings": []
    }
    
    # 1. Arithmetic: Total Call = Principal + Fees + Expenses
    calculated_call = calculated_call = (
    data.principal_call 
    + data.management_fee 
    + data.expenses 
    + data.equalization_interest 
    + data.recallable_capital 
    - abs(data.fee_offsets)
)
    if calculated_call != data.total_amount_due:
        diff = abs(calculated_call - data.total_amount_due)
        if diff <= Decimal("0.02"):
            audit_results["warnings"].append(f"Minor rounding variance detected: {diff} {data.currency}")
        else:
            audit_results["math_passed"] = False
            audit_results["errors"].append(
                f"Arithmetic Mismatch: Sum of items ({calculated_call}) != Total Stated ({data.total_amount_due})"
            )

    # 2. Arithmetic: Total Commitment Reconciliation Check
    reconciled_commitment = data.prior_contributed + data.principal_call + data.remaining_uncalled
    if data.total_commitment > 0 and reconciled_commitment != data.total_commitment:
        audit_results["warnings"].append(
            f"Commitment reconciliation offset: Expected {data.total_commitment}, Calculated {reconciled_commitment}"
        )

    # 3. Wire Validation: IBAN Checksum
    if data.iban:
        clean_iban = data.iban.replace(" ", "").upper()
        try:
            IBAN(clean_iban)
        except ValueError as e:
            audit_results["wire_passed"] = False
            audit_results["errors"].append(f"Invalid IBAN format/checksum: {str(e)}")

    # 4. Fraud Prevention: Baseline Mandate Verification
    if pre_approved_mandate and data.iban:
        clean_iban = data.iban.replace(" ", "").upper()
        expected_iban = pre_approved_mandate.get("iban", "").replace(" ", "").upper()
        
        if clean_iban != expected_iban:
            audit_results["wire_passed"] = False
            audit_results["errors"].append(
                f"CRITICAL WIRE ALERT: Notice IBAN ({clean_iban}) does NOT match approved baseline mandate ({expected_iban})."
            )

    return audit_results

File: src/lib/auditor.ts
import { NoticeExtraction, MandateRecord, AuditResult } from '../types';

/**
 * Validates international IBAN checksum using the standard ISO 7064 Modulo 97-10 algorithm (equivalent to Python schwifty IBAN)
 */
export function validateIBANChecksum(iban: string): { valid: boolean; error?: string } {
  if (!iban) return { valid: false, error: 'IBAN is required' };
  const clean = iban.replace(/[^A-Za-z0-9]/g, '').toUpperCase();
  
  if (clean.length < 15 || clean.length > 34) {
    return { valid: false, error: `Invalid IBAN character length (${clean.length})` };
  }

  // Country code format check (2 uppercase letters)
  const countryCode = clean.slice(0, 2);
  if (!/^[A-Z]{2}$/.test(countryCode)) {
    return { valid: false, error: 'Invalid country code prefix' };
  }

  // Move first 4 characters to the end
  const rearranged = clean.slice(4) + clean.slice(0, 4);

  // Convert letters to numbers (A=10, B=11, ..., Z=35)
  let numericString = '';
  for (let i = 0; i < rearranged.length; i++) {
    const code = rearranged.charCodeAt(i);
    if (code >= 65 && code <= 90) {
      numericString += (code - 55).toString();
    } else if (code >= 48 && code <= 57) {
      numericString += rearranged[i];
    } else {
      return { valid: false, error: 'Contains illegal characters' };
    }
  }

  // Perform modulo 97 in chunks to avoid numeric overflow
  let remainder = 0;
  for (let i = 0; i < numericString.length; i += 7) {
    const part = remainder.toString() + numericString.substring(i, i + 7);
    remainder = parseInt(part, 10) % 97;
  }

  if (remainder === 1) {
    return { valid: true };
  } else {
    return { valid: false, error: `Modulo 97 checksum failed (remainder = ${remainder}, expected 1)` };
  }
}

/**
 * Deterministic audit logic strictly matching Python auditor.py
 */
export function auditNotice(data: NoticeExtraction, preApprovedMandate?: MandateRecord | { fund_name?: string; iban?: string } | null): AuditResult {
  const auditResults: AuditResult = {
    math_passed: true,
    wire_passed: true,
    commitment_reconciled: true,
    calculated_call: 0,
    discrepancy_amount: 0,
    errors: [],
    warnings: [],
    iban_valid_checksum: true,
    mandate_matched: true,
  };

  // 1. Arithmetic: Total Call = Principal + Fees + Expenses + Equalization + Recallable - abs(Fee Offsets)
  const principal = Number(data.principal_call) || 0;
  const mgmtFee = Number(data.management_fee) || 0;
  const expenses = Number(data.expenses) || 0;
  const equalization = Number(data.equalization_interest) || 0;
  const recallable = Number(data.recallable_capital) || 0;
  const feeOffsets = Math.abs(Number(data.fee_offsets) || 0);

  const calculatedCall = Number((principal + mgmtFee + expenses + equalization + recallable - feeOffsets).toFixed(2));
  auditResults.calculated_call = calculatedCall;

  const totalStated = Number((Number(data.total_amount_due) || 0).toFixed(2));
  const diff = Number(Math.abs(calculatedCall - totalStated).toFixed(2));
  auditResults.discrepancy_amount = diff;

  if (calculatedCall !== totalStated) {
    if (diff <= 0.02) {
      auditResults.warnings.push(`Minor rounding variance detected: ${diff.toFixed(2)} ${data.currency}`);
    } else {
      auditResults.math_passed = false;
      auditResults.errors.push(
        `Arithmetic Mismatch: Sum of items (${calculatedCall.toLocaleString(undefined, { minimumFractionDigits: 2 })} ${data.currency}) != Total Stated (${totalStated.toLocaleString(undefined, { minimumFractionDigits: 2 })} ${data.currency})`
      );
    }
  }

  // 2. Arithmetic: Total Commitment Reconciliation Check
  const priorContributed = Number(data.prior_contributed) || 0;
  const remainingUncalled = Number(data.remaining_uncalled) || 0;
  const totalCommitment = Number(data.total_commitment) || 0;
  const reconciledCommitment = Number((priorContributed + principal + remainingUncalled).toFixed(2));

  if (totalCommitment > 0 && reconciledCommitment !== totalCommitment) {
    auditResults.commitment_reconciled = false;
    auditResults.warnings.push(
      `Commitment reconciliation offset: Expected ${totalCommitment.toLocaleString(undefined, { minimumFractionDigits: 2 })}, Calculated ${reconciledCommitment.toLocaleString(undefined, { minimumFractionDigits: 2 })}`
    );
  }

  // 3. Wire Validation: IBAN Checksum (using Modulo-97)
  if (data.iban) {
    const cleanIban = data.iban.replace(/\s+/g, '').toUpperCase();
    const ibanCheck = validateIBANChecksum(cleanIban);
    if (!ibanCheck.valid) {
      auditResults.wire_passed = false;
      auditResults.iban_valid_checksum = false;
      auditResults.errors.push(`Invalid IBAN format/checksum: ${ibanCheck.error || 'Invalid checksum'}`);
    }
  }

  // 4. Fraud Prevention: Baseline Mandate Verification
  if (preApprovedMandate && preApprovedMandate.iban && data.iban) {
    const cleanIban = data.iban.replace(/\s+/g, '').toUpperCase();
    const expectedIban = preApprovedMandate.iban.replace(/\s+/g, '').toUpperCase();

    if (expectedIban && cleanIban !== expectedIban) {
      auditResults.wire_passed = false;
      auditResults.mandate_matched = false;
      auditResults.errors.push(
        `CRITICAL WIRE ALERT: Notice IBAN (${cleanIban}) does NOT match approved baseline mandate (${expectedIban}).`
      );
    }
  }

  return auditResults;
}
